internal/core/src/index/BitmapIndex.cpp

```diff
@@ -0,0 +1,648 @@
+    std::vector<FieldDataPtr> field_datas;
+    for (auto rec = reader->Next(); rec != nullptr; rec = reader->Next()) {
+        if (!rec.ok()) {
+            PanicInfo(DataFormatBroken, "failed to read data");
+        }
+        auto data = rec.ValueUnsafe();
+        auto total_num_rows = data->num_rows();
+        auto col_data = data->GetColumnByName(field_name);
+        auto field_data = storage::CreateFieldData(
+            DataType(GetDType<T>()), 0, total_num_rows);
+        field_data->FillFieldData(col_data);
+        field_datas.push_back(field_data);
+    }
+
+    int total_num_rows = 0;
+    for (auto& field_data : field_datas) {
+        total_num_rows += field_data->get_num_rows();
+    }
+    if (total_num_rows == 0) {
+        throw SegcoreError(DataIsEmpty,
+                           "scalar bitmap index can not build null values");
+    }
+
+    total_num_rows_ = total_num_rows;
+
+    int64_t offset = 0;
+    for (const auto& data : field_datas) {
+        auto slice_row_num = data->get_num_rows();
+        for (size_t i = 0; i < slice_row_num; ++i) {
+            auto val = reinterpret_cast<const T*>(data->RawValue(i));
+            data_[*val].add(offset);
+            offset++;
+        }
+    }
+    is_built_ = true;
+}
+
+template <typename T>
+void
+BitmapIndex<T>::SerializeIndexData(uint8_t* data_ptr) {
+    for (auto& pair : data_) {
+        memcpy(data_ptr, &pair.first, sizeof(T));
+        data_ptr += sizeof(T);
+
+        pair.second.write(reinterpret_cast<char*>(data_ptr));
+        data_ptr += pair.second.getSizeInBytes();
+    }
+}
+
+template <>
+void
+BitmapIndex<std::string>::SerializeIndexData(uint8_t* data_ptr) {
+    for (auto& pair : data_) {
+        size_t key_size = pair.first.size();
+        memcpy(data_ptr, &key_size, sizeof(size_t));
+        data_ptr += sizeof(size_t);
+
+        memcpy(data_ptr, pair.first.data(), key_size);
+        data_ptr += key_size;
+
+        pair.second.write(reinterpret_cast<char*>(data_ptr));
+        data_ptr += pair.second.getSizeInBytes();
+    }
+}
+
+template <typename T>
+BinarySet
+BitmapIndex<T>::Serialize(const Config& config) {
+    AssertInfo(is_built_, "index has not been built yet");
+
+    auto index_data_size = 0;
+    for (auto& pair : data_) {
+        index_data_size += pair.second.getSizeInBytes() + sizeof(T);
+    }
+
+    std::shared_ptr<uint8_t[]> index_data(new uint8_t[index_data_size]);
+    uint8_t* data_ptr = index_data.get();
+    SerializeIndexData(data_ptr);
+
+    std::shared_ptr<uint8_t[]> index_length(new uint8_t[sizeof(size_t)]);
+    auto index_size = data_.size();
+    memcpy(index_length.get(), &index_size, sizeof(size_t));
+
+    std::shared_ptr<uint8_t[]> num_rows(new uint8_t[sizeof(size_t)]);
+    memcpy(num_rows.get(), &total_num_rows_, sizeof(size_t));
+
+    BinarySet ret_set;
+    ret_set.Append(BITMAP_INDEX_DATA, index_data, index_data_size);
+    ret_set.Append(BITMAP_INDEX_LENGTH, index_length, sizeof(size_t));
```

Comment:
These two index files are too small, which is not a good case for storing them on object storage. Maybe we can consider to combine them in only one file.

Reply:
It is good for code explain, but not good for store. for now, it just two files, this problem not very serious, if meta number is big, we may refactor this design, sum these meta to  a file.